### PR TITLE
Add Doxygen CI check and pre-commit hook

### PR DESCRIPTION
**Pull request by PointKernel**
This work is mainly taken from https://github.com/rapidsai/cudf/pull/11076.

It adds Doxygen CI check and pre-commit hook. All the existing Doxygen warnings are fixed.

### CODE REVIEW

File: doxygen/Doxyfile [hunk 1]
```diff
@@ -0,0 +1,2490 @@
...
+# the path). If a tag file is not located in the directory in which doxygen is
+# run, you must also specify the path to the tagfile here.
+
+TAGFILES               = rmm.tag=https://docs.rapids.ai/api/librmm/22.08
+
+# When a file name is specified after GENERATE_TAGFILE, doxygen will create a
+# tag file that is based on the input files it reads. See section "Linking to
+# external documentation" for more information about the usage of tag files.
+
+GENERATE_TAGFILE       = html/libcudf.tag
```
Comment (karthikeyann) [inline doxygen/Doxyfile]: name should be updated

File: doxygen/Doxyfile [hunk 2]
```diff
@@ -0,0 +1,2490 @@
...
+# Adding location for the tag files is done as follows:
+# TAGFILES = file1=loc1 "file2 = loc2" ...
+# where loc1 and loc2 can be relative or absolute paths or URLs. See the
+# section "Linking to external documentation" for more information about the use
+# of tag files.
+# Note: Each tag file must have a unique name (where the name does NOT include
+# the path). If a tag file is not located in the directory in which doxygen is
+# run, you must also specify the path to the tagfile here.
+
+TAGFILES               = rmm.tag=https://docs.rapids.ai/api/librmm/22.08
```
Comment (karthikeyann) [inline doxygen/Doxyfile]: is rmm used in this library?

File: doxygen/Doxyfile [hunk 3]
```diff
@@ -0,0 +1,2490 @@
...
+# If the value of the INPUT tag contains directories, you can use the
+# EXCLUDE_PATTERNS tag to specify one or more wildcard patterns to exclude
+# certain files from those directories.
+#
+# Note that the wildcards are matched against the file with absolute path, so to
+# exclude all test directories for example use the pattern */test/*
+
+EXCLUDE_PATTERNS       = */nvtx/* \
+                         */detail/* \
+                         */cudf_test/*
```
Comment (karthikeyann) [inline doxygen/Doxyfile:897]:
```suggestion
                         */detail/* 
```

File: doxygen/Doxyfile [hunk 4]
```diff
@@ -0,0 +1,2490 @@
...
+# created by doxygen. Using this option one can overrule certain style aspects.
+# This is preferred over using HTML_STYLESHEET since it does not replace the
+# standard style sheet and is therefore more robust against future updates.
+# Doxygen will copy the style sheet files to the output directory.
+# Note: The order of the extra style sheet files is of importance (e.g. the last
+# style sheet in the list overrules the setting of the previous ones in the
+# list). For an example see the documentation.
+# This tag requires that the tag GENERATE_HTML is set to YES.
+
+HTML_EXTRA_STYLESHEET  = rapids.css
```
Comment (karthikeyann) [inline doxygen/Doxyfile]: is this required? There are no custom pages which require this css in this repo.

**Review by karthikeyann — CHANGES_REQUESTED** [review]
Great work!
`ci/checks/style.sh` runs only clang-format. If you would like to run doxygen check in ci, _style.sh_ needs to be updated.

Also, is there a plan to publish doxygen  documentation in github.io pages?

File: doxygen/Doxyfile (hunk 3 above)
Comment (karthikeyann) [inline doxygen/Doxyfile]:
```suggestion
EXCLUDE_PATTERNS       = */detail/*
```

File: doxygen/Doxyfile (hunk 2 above)
Comment (PointKernel) [reply doxygen/Doxyfile]: Not at all. Me just copy-paste your work.

Removed.

**Comment by PointKernel** [comment]
> If you would like to run doxygen check in ci, style.sh needs to be updated.

I noticed in the cudf `style.sh` there is no command invoking `doxygen.sh` thus wonder how doxygen check is triggered in CI.

> is there a plan to publish doxygen documentation in github.io pages?

Nothing solid yet but yes, it's on my to-do list. Any comments here?

**Comment by karthikeyann** [comment]
> I noticed in the cudf `style.sh` there is no command invoking `doxygen.sh` thus wonder how doxygen check is triggered in CI.

https://github.com/rapidsai/cudf/blob/4bbb65876d1d775a8523e416fddae0d805caefb0/ci/checks/style.sh#L23
cudf runs all pre-commit hooks against all files. 
`pre-commit run --hook-stage manual --all-files`

> > is there a plan to publish doxygen documentation in github.io pages?
> 
> Nothing solid yet but yes, it's on my to-do list. Any comments here?

Depends on when documentation should be updated. I don't see releases or tags in this repo. 
How about updating documentation on every push to `main` branch? (last updated July 16, 2020 ! not update for long time!)

**Comment by PointKernel** [comment]
> cudf runs all pre-commit hooks against all files.

Great, I've updated the cuco script to do the same.

@karthikeyann Can you please re-review? Also, thanks for filing #178 to track the doc automation separately.

**Review by karthikeyann — CHANGES_REQUESTED** [review]
Fix for remaining doxygen warnings as patch. 
(I don't have write permissions to this repo, I could not push my commit to this PR)
```
From 63635ee3f967e82b444376b8eab4a56462ace388 Mon Sep 17 00:00:00 2001
From: Karthikeyan Natarajan <karthikeyann@users.noreply.github.com>
Date: Tue, 21 Jun 2022 10:47:20 +0530
Subject: [PATCH 1/1] fix some doxygen warnings

---
 include/cuco/detail/probe_sequence_impl.cuh | 27 ++++++++++++++-------
 1 file changed, 18 insertions(+), 9 deletions(-)

diff --git a/include/cuco/detail/probe_sequence_impl.cuh b/include/cuco/detail/probe_sequence_impl.cuh
index 6fdb32b..c7ad07b 100644
--- a/include/cuco/detail/probe_sequence_impl.cuh
+++ b/include/cuco/detail/probe_sequence_impl.cuh
@@ -47,7 +47,7 @@ class probe_sequence_base {
   static constexpr uint32_t vector_width() noexcept { return 2u; }
 };

-/*
+/**
  * @brief Base class of probe sequence implementation.
  *
  * Hash map operations are generally memory-bandwidth bound. A vector-load loads two consecutive
@@ -70,14 +70,17 @@ template <typename Key,
           uint32_t CGSize>
 class probe_sequence_impl_base {
  protected:
-  using value_type         = cuco::pair_type<Key, Value>;
-  using key_type           = Key;
-  using mapped_type        = Value;
-  using atomic_key_type    = cuda::atomic<key_type, Scope>;
-  using atomic_mapped_type = cuda::atomic<mapped_type, Scope>;
-  using pair_atomic_type   = cuco::pair_type<atomic_key_type, atomic_mapped_type>;
-  using iterator           = pair_atomic_type*;
-  using const_iterator     = pair_atomic_type const*;
+  using value_type         = cuco::pair_type<Key, Value>;       ///< Type of key/value pairs
+  using key_type           = Key;                               ///< Key type
+  using mapped_type        = Value;                             ///< Type of mapped values
+  using atomic_key_type    = cuda::atomic<key_type, Scope>;     ///< Type of atomic keys
+  using atomic_mapped_type = cuda::atomic<mapped_type, Scope>;  ///< Type of atomic mapped values
+  /// Pair type of atomic key and atomic mapped value
+  using pair_atomic_type = cuco::pair_type<atomic_key_type, atomic_mapped_type>;
+  /// Type of the forward iterator to `pair_atomic_type`
+  using iterator = pair_atomic_type*;
+  /// Type of the forward iterator to `const pair_atomic_type`
+  using const_iterator = pair_atomic_type const*;

   /**
    * @brief Returns the number of elements loaded with each vector-load.
@@ -115,6 +118,8 @@ class probe_sequence_impl_base {
  public:
   /**
    * @brief Returns the capacity of the hash map.
+   *
+   * @return The capacity of the hash map.
    */
   __host__ __device__ __forceinline__ std::size_t get_capacity() const noexcept
   {
@@ -123,11 +128,15 @@ class probe_sequence_impl_base {

   /**
    * @brief Returns slots array.
+   *
+   * @return Slots array
    */
   __device__ __forceinline__ iterator get_slots() noexcept { return slots_; }

   /**
    * @brief Returns slots array.
+   *
+   * @return Slots array
    */
   __device__ __forceinline__ const_iterator get_slots() const noexcept { return slots_; }

--
2.36.1

```

**Review by karthikeyann — APPROVED** [review]
LGTM 👍

File: include/cuco/detail/static_multimap/static_multimap.inl [hunk 5]
```diff
@@ -807,7 +807,7 @@ template <uint32_t buffer_size,
           typename OutputIt1,
           typename OutputIt2,
           typename PairEqual>
-__device__ __forceinline__ void
+__device__ inline void
```
Comment (jrhemstad) [inline include/cuco/detail/static_multimap/static_multimap.inl]: `__forceinline__` and `inline` are not the same. This change seems out of scope for this PR.

File: include/cuco/static_map.cuh [hunk 6]
```diff
@@ -384,9 +392,9 @@ class static_map {
    *
    * Writes a `bool` to `(output + i)` indicating if the key `*(first + i)` exists in the map.
    *
-   * Hash should be callable with both `std::iterator_traits<InputIt>::value_type` and Key type.
-   * `std::invoke_result<KeyEqual, std::iterator_traits<InputIt>::value_type, Key>` must be
-   * well-formed.
+   * Hash should be callable with both <tt>std::iterator_traits<InputIt>::value_type</tt> and Key
+   * type. <tt>std::invoke_result<KeyEqual, std::iterator_traits<InputIt>::value_type, Key></tt>
+   * must be well-formed.
```
Comment (jrhemstad) [inline include/cuco/static_map.cuh:397]: What's the deal with `<tt>`? afaik, markdown backticks work in Doxygen.

**Review by jrhemstad — CHANGES_REQUESTED** [review]
I'm excited for this!

I'd like to expand the Doxygen section of the README with an example of what our expectations/conventions are for documentation (we can steal this from libcudf).

The change from`__forceinline__` to `inline` seems orthogonal to this PR.

I also am not a fan of having to use `<tt></tt>` for code font. Whenever I've checked in the past, markdown ticks work in Doxygen.

File: include/cuco/detail/static_multimap/static_multimap.inl (hunk 5 above)
Comment (PointKernel) [reply include/cuco/detail/static_multimap/static_multimap.inl]: Doxygen cannot handle properly `__forceinline__` and it treats this keyword as a return type. So using it together with `void` will make Doxygen complain the doc of return type is missing. ~~I didn't find a way to silence this particular warning.~~ @karthikeyann may have a solution to this.

I noticed before that using `__forceinline__` has little performance impact compared to `inline` thus I was reverting back to `inline` to fix the Doxygen warning.

File: include/cuco/static_map.cuh (hunk 6 above)
Comment (PointKernel) [reply include/cuco/static_map.cuh:397]: Here's an example warning message of using backticks https://gpuci.gpuopenanalytics.com/blue/organizations/jenkins/nvidia%2Fgpuci%2FcuCollections%2Fprb%2FcuCollections-style/detail/cuCollections-style/147/pipeline#log-176. This happened exclusively to all docs containing the keyword "value_type".

Comment (jrhemstad) [reply include/cuco/static_map.cuh:397]: Which error in particular should I be looking at?

File: include/cuco/detail/static_multimap/static_multimap.inl (hunk 5 above)
Comment (jrhemstad) [reply include/cuco/detail/static_multimap/static_multimap.inl]: I'm surprised that the `__device__` annotation doesn't cause the same problem? 

Can we add an exception to ignore this warning for now? 

> I noticed before that using __forceinline__ has little performance impact compared to inline thus I was reverting back to inline to fix the Doxygen warning.

That's fine, but I'd like to tackle it as a separate PR where we can study and verify the perf changes in cudf as well.

Comment (PointKernel) [reply include/cuco/detail/static_multimap/static_multimap.inl]: > I'm surprised that the __device__ annotation doesn't cause the same problem?

`__device__` works just fine.

> Can we add an exception to ignore this warning for now?

Now I recall, we can silence this warning by turning off `WARN_NO_PARAMDOC` but I found this is an overkill since it will ignore undocumented parameters or return values which are basically all warnings we want to fix with Doxygen CI check.

File: include/cuco/static_map.cuh (hunk 6 above)
Comment (PointKernel) [reply include/cuco/static_map.cuh:397]: `warning: explicit link request to 'value_type' could not be resolved`

File: include/cuco/detail/static_multimap/static_multimap.inl (hunk 5 above)
Comment (PointKernel) [reply include/cuco/detail/static_multimap/static_multimap.inl]: Oh, I found the solution for this. We should add `__forceinline__` into `PREDEFINED` in the Doxygen config file to specify a custom macro.

File: include/cuco/static_map.cuh (hunk 6 above)
Comment (PointKernel) [reply include/cuco/static_map.cuh:397]: Here's what I found:
```cpp
* `std::iterator_traits<InputIt>`
```
Works fine but adding `::value_type` will fail with `explicit link request to 'value_type' could not be resolved`:
```cpp
* `std::iterator_traits<InputIt>::value_type`
```
I thought the problem is that we didn't provide a link to the implementation of "std::iterator_traits" thus the link request to the code fails. So maybe marking a locally defined function with backticks will pass? Following this idea, I tried the below code but had no luck either:
```cpp
* `cuco::detail::Murmurhash3_32<key_type>::result_type`
```

Another thing I tried is turning off `AUTOLINK_SUPPORT` in the doxygen config file. We can use backticks without getting any warnings by doing that but I think our intention is to keep the auto-link on in Doxygen. Any thoughts?

Comment (jrhemstad) [reply include/cuco/static_map.cuh:397]: I suspect it's expecting docs to exist for `std::iterator_traits<InputIt>::value_type` to exist and link to or something.

**Review by jrhemstad — APPROVED** [review]
